compiler/rustc_trait_selection/src/traits/util.rs

```diff
@@ -506,20 +507,80 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for PlaceholderReplacer<'_, 'tcx> {
     }
 }
 
-pub fn sizedness_fast_path<'tcx>(tcx: TyCtxt<'tcx>, predicate: ty::Predicate<'tcx>) -> bool {
-    // Proving `Sized` very often on "obviously sized" types like `&T`, accounts for about 60%
-    // percentage of the predicates we have to prove. No need to canonicalize and all that for
-    // such cases.
+pub fn sizedness_fast_path<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    predicate: ty::Predicate<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+) -> bool {
+    // Proving `Sized`/`MetaSized`/`PointeeSized`, very often on "obviously sized" types like
+    // `&T`, accounts for about 60% percentage of the predicates we have to prove. No need to
+    // canonicalize and all that for such cases.
     if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_ref)) =
         predicate.kind().skip_binder()
     {
-        if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized)
-            && trait_ref.self_ty().is_trivially_sized(tcx)
+        let sizedness = if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized) {
+            Some(SizedTraitKind::Sized)
+        } else if tcx.is_lang_item(trait_ref.def_id(), LangItem::MetaSized) {
+            Some(SizedTraitKind::MetaSized)
+        } else if tcx.is_lang_item(trait_ref.def_id(), LangItem::PointeeSized) {
+            Some(SizedTraitKind::PointeeSized)
+        } else {
+            None
```

Comment:
style: instead of using an `Option` here, early return if the trait isn't a `Sized` trait

Reply:
Fixed